core/node/da_dispatcher/src/da_dispatcher.rs

```diff
@@ -0,0 +1,147 @@
+use std::{future::Future, time::Duration};
+
+use anyhow::Error;
+use chrono::Utc;
+use tokio::sync::watch;
+use zksync_config::DADispatcherConfig;
+use zksync_da_layers::DataAvailabilityInterface;
+use zksync_dal::{Connection, ConnectionPool, Core, CoreDal};
+use zksync_types::L1BatchNumber;
+
+use crate::metrics::METRICS;
+
+#[derive(Debug)]
+pub struct DataAvailabilityDispatcher {
+    client: Box<dyn DataAvailabilityInterface>,
+    pool: ConnectionPool<Core>,
+    config: DADispatcherConfig,
+}
+
+impl DataAvailabilityDispatcher {
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        config: DADispatcherConfig,
+        client: Box<dyn DataAvailabilityInterface>,
+    ) -> Self {
+        Self {
+            pool,
+            config,
+            client,
+        }
+    }
+
+    pub async fn run(self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let pool = self.pool.clone();
+        loop {
+            let mut conn = pool.connection_tagged("da_dispatcher").await.unwrap();
+
+            if *stop_receiver.borrow() {
+                tracing::info!("Stop signal received, da_dispatcher is shutting down");
+                break;
+            }
+
+            if let Err(err) = self.dispatch(&mut conn).await {
+                tracing::warn!("dispatch error {err:?}");
+            }
+            if let Err(err) = self.poll_for_inclusion(&mut conn).await {
```

Comment:
Why is this causally ordered with `dispatch()` in this way? In other words, why doesn't polling for inclusion run concurrently and for multiple batches?

Reply:
it can be done concurrently because they are not dependent on each other, I just went with the simplest solution because I'm not sure more is needed here

polling for inclusion for multiple batches can also be done in parallel, but I don't see how it will be beneficial taking into account that we still submit all the batches sequentially(you can have a look into my response to Roman's comment above, there are more thoughts about this)